# An Examination of CAO Points Data 2019 - 2021
## Submission Assignment for Fundamentals of Data Analysis
## GMIT Higher Diploma in Data Analytics
#### Gerry Donnelly November 2021

### Introduction
This assignment is about getting data from websites and being able to analyse that data, all through Python. The specific use case in this instance is the points data contained on the CAO website. The objective of the assignment is to be able to automatically extract the relevant points data for the years 2019, 2020 and 2021 and complete an analysis of that data, comparing each of the years and assessing what are the key trends that emerge from the data. 

The CAO or Central Applications Office is the body that processes all applications for third level educational institutions that are part of the Irish Higher Education system (HEI's). Students sit the Leaving Certificate exam, i.e. the final assessment of second level education and on the basis of results in each subject taken are awarded points. The sum total of points achieved are used by the HEI's to make course offers to students on the basis of their overall points tally and their prior stated course preferences. All of the third level course offers to students are then made thorugh the CAO.

The CAO was instituted in 1977 and it is interesting to note that from a starting point of just 5 HEI's in 1977 and just under 15 thousand applicants it has grown steadily since then to reach 37 HEI's and in excess of 80 thousand applicants in 2021. The number of courses offered has gone from 69 in 1977 to over 1000 in 2021. [CAO Media Stats](http://www.cao.ie/index.php?page=mediapack&bb=mediastats)

While not considered in this assignment the growth, changes in the scale and course profiles offered by the CAO since its inception in 1977 provides an interesting insight into wider societal changes. Over this period Ireland has transformed itself from a largely rural and small scale industrial profile to one where some of the giants of 21st century global industry are happy to operate here. For example, one can only marvel at the scale of the site currently under construction at Intel in Leixlip, a scene unrecognisable to the leaving cert cohort of 1977 but entirely in keeping with the skillsets emerging from our 21st century education system.

On a narrower, but no less interesting time scale, i.e. 2019 - 2021, it will be interesting to see how the impact of a global pandemic and the complete upheaval of the education system during this period may be reflected in the CAO points trends over the three years. Apart from the mechanics of getting at the data, this will be the primary outcome of this assignment. 

### Assignment Outline

The assignment will be tackled in the following stages:
- Accessing and extracting the 2019 - 2021 points data from the CAO Website
- Creating a single pandas dataframe to hold all of the data.
- Use of the pandas dataframe to analyse the points data across the 3 years. 
- Use of plots to demonstrate the essential elements of the analysis.

In terms of assignment scope in relation to the data following should be noted:
- The analsyis will be confined to the points data for first round offers for **Level 8** courses only. While it is recognised that Levels 6 and 7 courses play an important role in higher education the clearest indicators of pandemic impacts are at the Level 8 courses. 
- To provide meaningful year on year comparisons the analysis will also be restricted to those courses that appear in all three years, the logic being this is the best way to see the year on year trends. I will use 2019 as the base year. 


#### Section 1 - Getting the Data. 
All the CAO points data is contained on the CAO website with links for each year going back to 1998. 

Initial Observations on the points Data
- For some reason the CAO do not appear to maintain a consistent approach to providing the annual points data. This provides some interesting challenges to get the data.
- Points 2019. The data is contained in a PDF document which requires additional steps to convert to a form useable in a pandas dataframe. 
- Points 2020. The data is contained in a Excel .xlsx format which can be used directly to import to a dataframe.  
- Points 2021. The data is now contained in an Excel .xlsx file, previously it was contained only in a web page.

However there are some common characteristics to the data common across all years that will form an important part of the extraction, transformations and analysis:
- Every Level 8 course is assigned a 5 character code, example *DC120 - Computing for Business.*
    - The first 2 characters are alphabetic and indicate the college, in the above example DC refers to Dublin City University.
    - The three digits indicate the course.
- In general the datasets contain course codes, course names, college names and points for each course, being initial offers and subsequect offer rounds. 

In addition all courses are alos assigned to an ISCED classification. [Link to ISCED Description](https://en.wikipedia.org/wiki/International_Standard_Classification_of_Education). This classification system, designed in the early 1970's and substantially updated in 1997 and 2011 serves to provide a standard international classification system across all educational levels in all counrties. It allows for course and course level comparibility within countries and internationally. The ISCED 2011 has been adopted across the EU since 2014 and the revised classification of fields of education and training ISCED-F-2013 has been implemented on education systems data collection since 2014/2015. [Link to ISCED Eurostat](https://ec.europa.eu/eurostat/statistics-explained/index.php?title=International_Standard_Classification_of_Education_(ISCED)#Implementation_of_ISCED_2011_.28levels_of_education.29).

For example in ISCED-F 2013 **06** defines the field of Information and Communication Technologies and in the CAO system 061 defines courses allocated to the general subject area of Information and Communication Technologies (ICTs). 

So for a full course classifications example CK411 is a Level 8 course in Data Science and Analytics course offered by University College Cork and is assigned to (061) Information and Communication Technologies (ICTs) for general classification. These classifications will be useful in looking at trends in general areas as against specific courses. 

Another point to note is that the points data also inlcudes additional characters that have to be dealt with, usually appended before or after the numeric digits. It also inlcudes non numeric data.
- The # character signifies that entry to the associated course is also dependant on an additional Test, Interview, Portfolio or Audition, e.g. #700 or #+matric.
- AQA indicates that the course had sufficient places to offer to all qualified applicants so no points data is provided. 
- The * character indicates that not all applicants achieveing the points score were offered places, final selection usually by lottery. 
- n/a characters, likely to indicate that there were no applicants or the course was not offered. 

For the purpose of the assignment I will start with the 2021 data and work backwards. Overall my objective is to get each of the years data and combine all to one data frame. The basic structure of this final dataset will be course code, course name, course category, college name, points, year. 

##### 2021 CAO Points data
As stated above the 2021 points data can be found at http://www.cao.ie/index.php?page=points&p=2021. 

Initial inspection of the file shows that the row headers are in row 12 of the file so when importing will ignore rows 1 - 11. 

In [235]:
# Import the regex library, might need this. 
import re

import pandas as pd

# Import the web requests library
import requests as rq

# Import the datetime library, will use for timestamping files
import datetime as dt

# Import os library, will use this to generate folder paths for saving files. 
import os



In [236]:
# Set up the url for the 2020 points data.It is already in excel format.
cao2021url = 'http://www2.cao.ie/points/CAOPointsCharts2021.xlsx'

In [237]:
# Read in the excel file for the 2021 points data, ignoring rows 1 - 11, get a first look at the data. 
df2021 = pd.read_excel(cao2021url, skiprows=11, usecols='A:O')
df2021.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1451 entries, 0 to 1450
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   CATEGORY (ISCED Description)  1451 non-null   object
 1   Course Title                  1451 non-null   object
 2   Course Code                   1451 non-null   object
 3   R1 Points                     1416 non-null   object
 4   R1 Random                     83 non-null     object
 5   R2 Points                     384 non-null    object
 6   R2 Random                     52 non-null     object
 7   EOS Points                    1422 non-null   object
 8   EOS Random                    74 non-null     object
 9   EOS Midpoints                 1433 non-null   object
 10  Course Level                  1451 non-null   int64 
 11  HEI                           1451 non-null   object
 12  Test/Interview                119 non-null    object
 13  AVP               

The above information describes the dataframe, some initial observations:
- All the data types except for Course level are are of type object, indicating either text or a mixture of text and numeric.
- Will need to get the R1 Points into a numerical object type. 
- While there are 1451 records there are only 1416 non-null records in the R1 points, will have to see what these are. 

In [238]:
# Apart form that the data looks ok, the 1451 rows and 15 columns matches with what is in the excel file. 
df2021

CATEGORY (ISCED Description)  \
0                    Engineering and engineering trades   
1                                                Health   
2                                                Health   
3                       Biological and related sciences   
4                           Business and administration   
...                                                 ...   
1446  Information and Communication Technologies (ICTs)   
1447                                  Personal services   
1448                 Engineering and engineering trades   
1449                                            Welfare   
1450                        Business and administration   

                                  Course Title Course Code R1 Points  \
0              Music and Instrument Technology       AL605       211   
1                          Pharmacy Technician       AL630       308   
2                               Dental Nursing       AL631       311   
3                              Applied Science       AL632       297   
4                                     Business       AL650       AQA   
...                                        ...         ...       ...   
1446                        Creative Computing       WD211       270   
1447           Recreation and Sport Management       WD212       262   
1448  Mechanical and Manufacturing Engineering       WD230       230   
1449        Early Childhood Care and Education       WD231       266   
1450              Business Information Systems       WD232       261   

     R1 Random R2 Points  R2 Random EOS Points EOS Random EOS Midpoints  \
0          NaN        NaN       NaN        211        NaN           319   
1          NaN        NaN       NaN        308        NaN           409   
2          NaN        NaN       NaN        311        NaN           400   
3          NaN        NaN       NaN        297        NaN           454   
4          NaN        AQA       NaN        AQA        NaN           351   
...        ...        ...       ...        ...        ...           ...   
1446       NaN        NaN       NaN        270        NaN           392   
1447       NaN        NaN       NaN        262        NaN           304   
1448       NaN        230       NaN        230        NaN           361   
1449       NaN        NaN       NaN        266        NaN           366   
1450       NaN        261       NaN        261        NaN           329   

      Course Level                                HEI Test/Interview  AVP    v  
0                6    Athlone Institute of Technology            NaN  NaN  NaN  
1                6    Athlone Institute of Technology            NaN  NaN  NaN  
2                6    Athlone Institute of Technology            NaN  NaN  NaN  
3                6    Athlone Institute of Technology            NaN  NaN  NaN  
4                6    Athlone Institute of Technology            NaN  avp  NaN  
...            ...                                ...            ...  ...  ...  
1446             8  Waterford Institute of Technology            NaN  NaN  NaN  
1447             8  Waterford Institute of Technology            NaN  NaN  NaN  
1448             8  Waterford Institute of Technology            NaN  avp  NaN  
1449             8  Waterford Institute of Technology            NaN  NaN  NaN  
1450             8  Waterford Institute of Technology            NaN  avp  NaN  

[1451 rows x 15 columns]

In [239]:
# Save the original file.
df2021.to_csv('./CAO_Data/cao21_orig.csv',index=False)

In [240]:
# Next step is to create a Level 8 file only, so will extract the Level 8 courses only using the Course Level column. 
df2021_lvl8 = df2021[df2021['Course Level']==8]
df2021_lvl8.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1034 entries, 29 to 1450
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   CATEGORY (ISCED Description)  1034 non-null   object
 1   Course Title                  1034 non-null   object
 2   Course Code                   1034 non-null   object
 3   R1 Points                     1007 non-null   object
 4   R1 Random                     82 non-null     object
 5   R2 Points                     261 non-null    object
 6   R2 Random                     51 non-null     object
 7   EOS Points                    1012 non-null   object
 8   EOS Random                    74 non-null     object
 9   EOS Midpoints                 1016 non-null   object
 10  Course Level                  1034 non-null   int64 
 11  HEI                           1034 non-null   object
 12  Test/Interview                99 non-null     object
 13  AVP              

Now we have the level 8 courses only, looks like 27 of the R1 points are null. Good news is that the essential columns; Category, Course Title, Course Code, Course Level and HEI all have 1034 records. 

In [241]:
# Have a look at the courses that have no R1 Points data, not sure why so, maybe no applicants, anyway these will be excluded from the analysis. 
nullpoints_21 = pd.isnull(df2021_lvl8['R1 Points'])
df2021_lvl8[nullpoints_21]

CATEGORY (ISCED Description)  \
34                        Architecture and construction   
67    Information and Communication Technologies (ICTs)   
221                                             Welfare   
287   Information and Communication Technologies (ICTs)   
289   Information and Communication Technologies (ICTs)   
290   Information and Communication Technologies (ICTs)   
404                          Journalism and information   
406                         Business and administration   
412                         Business and administration   
432                         Business and administration   
433                         Business and administration   
458                                              Health   
723                                                Arts   
732                                                Arts   
737                                                Arts   
747                                                Arts   
753                                                Arts   
857                        Manufacturing and processing   
950                                             Welfare   
958                                                Arts   
1222  Interdisciplinary programmes and qualification...   
1224                                          Education   
1257                                            Welfare   
1258                    Social and behavioural sciences   
1317                 Engineering and engineering trades   
1369                      Architecture and construction   
1383                                            Welfare   

                                           Course Title Course Code R1 Points  \
34                                    Civil Engineering       AL811       NaN   
67            Creative Computing and Digital Innovation       CW048       NaN   
221                Social Work - Mature Applicants only       CK115       NaN   
287                         Computing (Cloud Computing)       DB501       NaN   
289                    Computing (Software Development)       DB503       NaN   
290               Computing (Web and Mobile Technology)       DB504       NaN   
404                 Communications and Media Production       GC250       NaN   
406                              Accounting and Finance       GC301       NaN   
412                                      Business (HRM)       GC402       NaN   
432                              Accounting and Finance       ID001       NaN   
433                                    Business Studies       ID002       NaN   
458         Medicine - Graduate Entry (GAMSAT required)       RC101       NaN   
723                                    History/Religion       TR454       NaN   
732   Middle Eastern,Jewish and Islamic Civilisation...       TR564       NaN   
737   Middle Eastern,Jewish and Islamic Civilisation...       TR587       NaN   
747                                      Music/Religion       TR638       NaN   
753                            Religion/Modern Language       TR666       NaN   
857                                Agri-Food Production       DK885       NaN   
950      Community Development and Youth Work (Level 8)       GA892       NaN   
958                 Furniture and Digital Manufacturing       GA983       NaN   
1222                                           Theology       MU002       NaN   
1224  Education - Primary Teaching - Gaeltacht Appli...       MH002       NaN   
1257   Community and Youth Work (part-time, in service)       MH802       NaN   
1258  Local Studies/Community Studies (part-time, ev...       MH803       NaN   
1317          Electronics and Self Driving Technologies       SG349       NaN   
1369                            Construction Management       TL842       NaN   
1383  Counselling with Addiction - Mature applicants...       TL889       NaN   

     R1 Random R2 Points  R2 Random EOS Points EOS Random EOS Midpoints  \
34         NaN        NaN     

In [242]:
# Get only the courses that have data in the R1 points column, will have a look further to see if they all have valid points data.
notnullr1 = pd.notnull(df2021_lvl8['R1 Points'])
df2021_lvl8 = df2021_lvl8[notnullr1]
df2021_lvl8

CATEGORY (ISCED Description)  \
29    Information and Communication Technologies (ICTs)   
30    Information and Communication Technologies (ICTs)   
31    Information and Communication Technologies (ICTs)   
32    Information and Communication Technologies (ICTs)   
33                        Architecture and construction   
...                                                 ...   
1446  Information and Communication Technologies (ICTs)   
1447                                  Personal services   
1448                 Engineering and engineering trades   
1449                                            Welfare   
1450                        Business and administration   

                                           Course Title Course Code R1 Points  \
29       Software Design for Virtual Reality and Gaming       AL801       300   
30    Software Design in Artificial Intelligence for...       AL802       313   
31    Software Design for Mobile Apps and Connected ...       AL803       350   
32      Computer Engineering for Network Infrastructure       AL805       321   
33                                   Quantity Surveying       AL810       328   
...                                                 ...         ...       ...   
1446                                 Creative Computing       WD211       270   
1447                    Recreation and Sport Management       WD212       262   
1448           Mechanical and Manufacturing Engineering       WD230       230   
1449                 Early Childhood Care and Education       WD231       266   
1450                       Business Information Systems       WD232       261   

     R1 Random R2 Points  R2 Random EOS Points EOS Random EOS Midpoints  \
29         NaN        NaN       NaN        300        NaN           359   
30         NaN        NaN       NaN        313        NaN           381   
31         NaN        NaN       NaN        350        NaN           398   
32         NaN        NaN       NaN        321        NaN           381   
33         NaN        NaN       NaN        328        NaN           377   
...        ...        ...       ...        ...        ...           ...   
1446       NaN        NaN       NaN        270        NaN           392   
1447       NaN        NaN       NaN        262        NaN           304   
1448       NaN        230       NaN        230        NaN           361   
1449       NaN        NaN       NaN        266        NaN           366   
1450       NaN        261       NaN        261        NaN           329   

      Course Level                                HEI Test/Interview  AVP    v  
29               8    Athlone Institute of Technology            NaN  NaN  NaN  
30               8    Athlone Institute of Technology            NaN  NaN  NaN  
31               8    Athlone Institute of Technology            NaN  NaN  NaN  
32               8    Athlone Institute of Technology            NaN  NaN  NaN  
33               8    Athlone Institute of Technology            NaN  NaN  NaN  
...            ...                                ...            ...  ...  ...  
1446             8  Waterford Institute of Technology            NaN  NaN  NaN  
1447             8  Waterford Institute of Technology            NaN  NaN  NaN  
1448             8  Waterford Institute of Technology            NaN  avp  NaN  
1449             8  Waterford Institute of Technology            NaN  NaN  NaN  
1450             8  Waterford Institute of Technology            NaN  avp  NaN  

[1007 rows x 15 columns]

In [243]:
# Rename the ISCED col.
df2021_lvl8=df2021_lvl8.rename(columns={'CATEGORY (ISCED Description)': 'ISCED'})

In [244]:
# Put in a HEI Code
df2021_lvl8['HEI Code'] = df2021_lvl8['Course Code'].str[:2]

In [245]:
# Pick out the columns needed for the analysis. 
df2021_lvl8 = df2021_lvl8[['Course Code', 'Course Title', 'HEI Code', 'HEI', 'ISCED', 'R1 Points']]

In [246]:
# Add the Year.
df2021_lvl8['Year']=str(2021)

In [247]:
df2021_lvl8

Course Code                                       Course Title HEI Code  \
29         AL801     Software Design for Virtual Reality and Gaming       AL   
30         AL802  Software Design in Artificial Intelligence for...       AL   
31         AL803  Software Design for Mobile Apps and Connected ...       AL   
32         AL805    Computer Engineering for Network Infrastructure       AL   
33         AL810                                 Quantity Surveying       AL   
...          ...                                                ...      ...   
1446       WD211                                 Creative Computing       WD   
1447       WD212                    Recreation and Sport Management       WD   
1448       WD230           Mechanical and Manufacturing Engineering       WD   
1449       WD231                 Early Childhood Care and Education       WD   
1450       WD232                       Business Information Systems       WD   

                                    HEI  \
29      Athlone Institute of Technology   
30      Athlone Institute of Technology   
31      Athlone Institute of Technology   
32      Athlone Institute of Technology   
33      Athlone Institute of Technology   
...                                 ...   
1446  Waterford Institute of Technology   
1447  Waterford Institute of Technology   
1448  Waterford Institute of Technology   
1449  Waterford Institute of Technology   
1450  Waterford Institute of Technology   

                                                  ISCED R1 Points  Year  
29    Information and Communication Technologies (ICTs)       300  2021  
30    Information and Communication Technologies (ICTs)       313  2021  
31    Information and Communication Technologies (ICTs)       350  2021  
32    Information and Communication Technologies (ICTs)       321  2021  
33                        Architecture and construction       328  2021  
...                                                 ...       ...   ...  
1446  Information and Communication Technologies (ICTs)       270  2021  
1447                                  Personal services       262  2021  
1448                 Engineering and engineering trades       230  2021  
1449                                            Welfare       266  2021  
1450                        Business and administration       261  2021  

[1007 rows x 7 columns]

In [248]:
# Check the R1 points, pull out any that dont begin with and contain only numbers from 0 - 9.
#df2021_lvl8[df2021_lvl8['R1 Points'].str.contains('^\d')==False]


In [249]:
# Use os.cwd to generate the current working directory for saving the output files from the analysis.
cwd = os.getcwd()
print(cwd)

C:\Users\donne\OneDrive\Documents\GMIT Data\Fundamentals of Data Analysis\Assessment\data_analysis_assessment


In [250]:
now = dt.datetime.now()
path = cwd + '/CAO_Data'+'/cao2021_' + now.strftime('%Y%m%d_%H%M%S')+'.csv'
#path = cwd + '/CAO_Data'+'/cao2021.csv'
print(path)

C:\Users\donne\OneDrive\Documents\GMIT Data\Fundamentals of Data Analysis\Assessment\data_analysis_assessment/CAO_Data/cao2021_20211201_073347.csv


In [251]:
# Save the file for 2021. 
df2021_lvl8.to_csv(path, index=False)

***

### CAO Points 2020

http://www2.cao.ie/points/CAOPointsCharts2020.xlsx

In [252]:
# Set up the url for the 2020 points data.It is already in excel formst.
cao2020url = 'http://www2.cao.ie/points/CAOPointsCharts2020.xlsx'

In [253]:
# Read in the excel file for the 2020 points data. In this file the header row in in row 11 so skip the first 10 rows. 
df2020 = pd.read_excel(cao2020url, skiprows=10, usecols='A:O')
df2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1464 entries, 0 to 1463
Data columns (total 15 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   CATEGORY (i.e.ISCED description)  1464 non-null   object
 1   COURSE TITLE                      1464 non-null   object
 2   COURSE CODE2                      1464 non-null   object
 3   R1 POINTS                         1437 non-null   object
 4   R1 Random *                       59 non-null     object
 5   R2 POINTS                         333 non-null    object
 6   R2 Random*                        29 non-null     object
 7   EOS                               1448 non-null   object
 8   EOS Random *                      69 non-null     object
 9   EOS Mid-point                     1437 non-null   object
 10  LEVEL                             1464 non-null   int64 
 11  HEI                               1464 non-null   object
 12  Test/Interview #    

Layout and object types same as 2021 data, will proceed in same way to create the dataframe needed. 

In [254]:
# Save the original file.
df2020.to_csv('./CAO_Data/cao20_orig.csv',index=False)

In [255]:
#Get and check the level 8 courses only. 
df2020_lvl8 = df2020[df2020['LEVEL']==8]
df2020_lvl8.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1027 entries, 0 to 1463
Data columns (total 15 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   CATEGORY (i.e.ISCED description)  1027 non-null   object
 1   COURSE TITLE                      1027 non-null   object
 2   COURSE CODE2                      1027 non-null   object
 3   R1 POINTS                         1008 non-null   object
 4   R1 Random *                       58 non-null     object
 5   R2 POINTS                         212 non-null    object
 6   R2 Random*                        29 non-null     object
 7   EOS                               1014 non-null   object
 8   EOS Random *                      69 non-null     object
 9   EOS Mid-point                     1003 non-null   object
 10  LEVEL                             1027 non-null   int64 
 11  HEI                               1027 non-null   object
 12  Test/Interview #    

In [256]:
# Have a look at the 2020 courses that have no R1 Points data, not sure why so, maybe no applicants, anyway these will be excluded from the analysis. 
nullpoints_20 = pd.isnull(df2020_lvl8['R1 POINTS'])
# There are 19 courses with no points. 
df2020_lvl8[nullpoints_20]

CATEGORY (i.e.ISCED description)  \
97                                              Welfare   
98                                              Welfare   
151                                           Education   
309   Information and Communication Technologies (ICTs)   
314                         Business and administration   
447                                         Agriculture   
448                        Manufacturing and processing   
487                                          Veterinary   
490                                              Health   
492                                              Health   
538                                           Education   
613                          Journalism and information   
616                         Business and administration   
631                                                Arts   
891                                           Education   
961                                              Health   
1073                                            Welfare   
1178                                               Arts   
1192                                               Arts   

                                           COURSE TITLE COURSE CODE2  \
97    Social Science (Youth and Community Work) - 3 ...        CK114   
98                 Social Work - Mature Applicants only        CK115   
151   Education - Primary Teaching - Gaeltacht Appli...        CM002   
309               Computing (Web and Mobile Technology)        DB504   
314                                  Financial Services        DB516   
447                             Sustainable Agriculture        DK884   
448                                Agri-Food Production        DK885   
487   Veterinary Medicine - Graduate Entry (GAMSAT r...        DN301   
490         Medicine - Graduate Entry (GAMSAT required)        DN401   
492                        Radiography - Graduate Entry        DN411   
538                Education (Art, Design and Graphics)        GA285   
613                 Communications and Media Production        GC250   
616                              Business Studies - HRM        GC302   
631                    Design Communication (portfolio)        GC462   
891   Education - Primary Teaching - Gaeltacht Appli...        MH002   
961         Medicine - Graduate Entry (GAMSAT required)        RC101   
1073  Counselling with Addiction - Mature applicants...        TL889   
1178  Middle Eastern,Jewish and Islamic Civilisation...        TR564   
1192                           Religion/Modern Language        TR666   

     R1 POINTS R1 Random * R2 POINTS R2 Random*  EOS EOS Random *  \
97         NaN         NaN       NaN        NaN  777          NaN   
98         NaN         NaN       NaN        NaN  999          NaN   
151        NaN         NaN       NaN        NaN  NaN          NaN   
309        NaN         NaN       281        NaN  281          NaN   
314        NaN         NaN       NaN        NaN  NaN          NaN   
447        NaN         NaN       NaN        NaN  NaN          NaN   
448        NaN         NaN       NaN        NaN  NaN          NaN   
487        NaN         NaN       NaN        NaN    #          NaN   
490        NaN         NaN       NaN        NaN   58            *   
492        NaN         NaN       NaN        NaN    #          NaN   
538        NaN         NaN       NaN        NaN  NaN          NaN   
613        NaN         NaN       NaN        NaN  NaN          NaN   
616        NaN         NaN       NaN        NaN  NaN          NaN   
631        NaN         NaN       NaN        NaN  NaN          NaN   
891        NaN         NaN       NaN        NaN  NaN          NaN   
961        NaN         NaN       NaN        NaN   61          NaN   
1073       NaN         NaN       NaN        NaN  NaN          NaN   
1178       NaN         NaN       NaN        NaN  NaN          NaN   
1192       NaN         NaN       NaN        NaN  NaN          NaN   

     EOS Mid-point  

In [257]:
# Get only the 2020 courses that have data in the R1 points column, will have a look further to see if they are all valid points data.
notnullr1 = pd.notnull(df2020_lvl8['R1 POINTS'])
df2020_lvl8 = df2020_lvl8[notnullr1]
df2020_lvl8

CATEGORY (i.e.ISCED description)  \
0                           Business and administration   
1                         Humanities (except languages)   
2                                                  Arts   
3                                                  Arts   
4                                                  Arts   
...                                                 ...   
1455                                               Arts   
1460  Information and Communication Technologies (ICTs)   
1461  Information and Communication Technologies (ICTs)   
1462                                  Personal services   
1463                 Engineering and engineering trades   

                                           COURSE TITLE COURSE CODE2  \
0                                International Business        AC120   
1                                          Liberal Arts        AC137   
2      First Year Art & Design (Common Entry,portfolio)        AD101   
3     Graphic Design and Moving Image Design (portfo...        AD102   
4     Textile & Surface Design and Jewellery & Objec...        AD103   
...                                                 ...          ...   
1455                                     Arts (options)        WD200   
1460                       Software Systems Development        WD210   
1461                                 Creative Computing        WD211   
1462                    Recreation and Sport Management        WD212   
1463           Mechanical and Manufacturing Engineering        WD230   

     R1 POINTS R1 Random * R2 POINTS R2 Random*       EOS EOS Random *  \
0          209         NaN       NaN        NaN       209          NaN   
1          252         NaN       NaN        NaN       252          NaN   
2     #+matric         NaN       NaN        NaN  #+matric          NaN   
3     #+matric         NaN       NaN        NaN  #+matric          NaN   
4     #+matric         NaN       NaN        NaN  #+matric          NaN   
...        ...         ...       ...        ...       ...          ...   
1455       AQA         NaN       AQA        NaN       AQA          NaN   
1460       279         NaN       NaN        NaN       279          NaN   
1461       271         NaN       NaN        NaN       271          NaN   
1462       270         NaN       NaN        NaN       270          NaN   
1463       253         NaN       NaN        NaN       253          NaN   

     EOS Mid-point  LEVEL                                 HEI  \
0              280      8                    American College   
1              270      8                    American College   
2         #+matric      8  National College of Art and Design   
3         #+matric      8  National College of Art and Design   
4         #+matric      8  National College of Art and Design   
...            ...    ...                                 ...   
1455           336      8   Waterford Institute of Technology   
1460           337      8   Waterford Institute of Technology   
1461           318      8   Waterford Institute of Technology   
1462           349      8   Waterford Institute of Technology   
1463           369      8   Waterford Institute of Technology   

     Test/Interview #  avp    v  
0                 NaN  NaN  NaN  
1                 NaN  NaN  NaN  
2                   #  NaN  NaN  
3                   #  NaN  NaN  
4                   #  NaN  NaN  
...               ...  ...  ...  
1455              NaN  avp  NaN  
1460              NaN  NaN  NaN  
1461              NaN  NaN  NaN  
1462              NaN  NaN  NaN  
1463              NaN  NaN  NaN  

[1008 rows x 15 columns]

In [258]:
# Rename the Columns to the standard list. 
df2020_lvl8=df2020_lvl8.rename(columns={'COURSE CODE2':'Course Code', 'COURSE TITLE':'Course Title','CATEGORY (i.e.ISCED description)': 'ISCED', 'R1 POINTS':'R1 Points'})

In [259]:
# Put in a HEI Code, will need that later for the 2019 data. 
df2020_lvl8['HEI Code'] = df2020_lvl8['Course Code'].str[:2]

In [260]:
# Pick out the columns needed for the analysis. 
df2020_lvl8 = df2020_lvl8[['Course Code', 'Course Title', 'HEI Code', 'HEI', 'ISCED', 'R1 Points']]

In [261]:
# Add the Year.
df2020_lvl8['Year']=str(2020)
df2020_lvl8

Course Code                                       Course Title HEI Code  \
0          AC120                             International Business       AC   
1          AC137                                       Liberal Arts       AC   
2          AD101   First Year Art & Design (Common Entry,portfolio)       AD   
3          AD102  Graphic Design and Moving Image Design (portfo...       AD   
4          AD103  Textile & Surface Design and Jewellery & Objec...       AD   
...          ...                                                ...      ...   
1455       WD200                                     Arts (options)       WD   
1460       WD210                       Software Systems Development       WD   
1461       WD211                                 Creative Computing       WD   
1462       WD212                    Recreation and Sport Management       WD   
1463       WD230           Mechanical and Manufacturing Engineering       WD   

                                     HEI  \
0                       American College   
1                       American College   
2     National College of Art and Design   
3     National College of Art and Design   
4     National College of Art and Design   
...                                  ...   
1455   Waterford Institute of Technology   
1460   Waterford Institute of Technology   
1461   Waterford Institute of Technology   
1462   Waterford Institute of Technology   
1463   Waterford Institute of Technology   

                                                  ISCED R1 Points  Year  
0                           Business and administration       209  2020  
1                         Humanities (except languages)       252  2020  
2                                                  Arts  #+matric  2020  
3                                                  Arts  #+matric  2020  
4                                                  Arts  #+matric  2020  
...                                                 ...       ...   ...  
1455                                               Arts       AQA  2020  
1460  Information and Communication Technologies (ICTs)       279  2020  
1461  Information and Communication Technologies (ICTs)       271  2020  
1462                                  Personal services       270  2020  
1463                 Engineering and engineering trades       253  2020  

[1008 rows x 7 columns]

In [262]:
# Check the R1 points, pull out any that dont begin with and contain only numbers from 0 - 9.
# df2020lvl8[df2020lvl8['R1 POINTS'].str.contains('^\d')==False]

In [263]:
# Save the file for 2020 Points.
path = cwd + '/CAO_Data'+'/cao2020_' + now.strftime('%Y%m%d_%H%M%S')+'.csv'
df2020_lvl8.to_csv(path, index=False)

### CAO Points 2019

In [264]:
# Set up the url for the 2019 points data, note it is a .pdf file so additional steps will be needed to get at the data. 
cao2019url = 'http://www2.cao.ie/points/lvl8_19.pdf'

Steps used to extract the pdf points data from the 2019 CAO file.
There are a number of ways to turn the pdf file into excel. To save time following was done:
- Open the file from the link above in Chrome browser.
- The built in Adobe Acrobat Chrome extension has a selection of file conversion tools, one of which is pdf to excel conversion.
- The resulting excel file opens in the browser and can be downloaded to a system folder, in this case to a folder named CAO_Data.
- The file is structured in sections, with a spection for each HEI, however the HEI information does not appear at the individual row level for each course but as a header for the section. 
- I have deleted the HEI header rows directly in the excel file.
- The file also does not contain any ISCED data. 

On further examination of the file that has been downloaded there are 111 courses that do not appear in either 2020 or 2021, they relate to courses with college code DT (Technological University Dublin City Campus), BN (Technological University Dublin Blanchardstown Campus) and TA (Technological University Dublin Tallaght Campus). In 2020 and 2021 the courses under these college codes appear to have been amalgamated into a common college code of TU, Technological University Dublin. However there is no relationship between the original 2019 Course Codes and the New 2020 and 2021 Course Codes. 

In [265]:
# Read in the 2019 points excel file, note this file contains only Level 8 courses. 
# The file also is missing the row level details for the HEI Name and ISCED Category.
# The R1 points data is in a columns named EOS.
# Create a College Code from the first 2 letters of the Course Code, can use this to get the College Name. 
df2019_lvl8 = pd.read_excel('lvl8_2019.xlsx')
df2019_lvl8.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 930 entries, 0 to 929
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Course Code  930 non-null    object 
 1   Course       930 non-null    object 
 2   EOS          926 non-null    object 
 3   Mid          915 non-null    float64
dtypes: float64(1), object(3)
memory usage: 29.2+ KB


In [266]:
# Have a look at the courses with null R1 values. There are 4 of them. 
nullpoints_19 = pd.isnull(df2019_lvl8['EOS'])
df2019_lvl8[nullpoints_19]

Course Code                         Course  EOS  Mid
366       GC462  Design Communication (Dublin)  NaN  NaN
371       ID003  Business Studies with Chinese  NaN  NaN
877       TL842        Construction Management  NaN  NaN
880       TL847      Manufacturing Engineering  NaN  NaN

In [267]:
# Remove the null value courses.
notnullr1 = pd.notnull(df2019_lvl8['EOS'])
df2019_lvl8 = df2019_lvl8[notnullr1]
df2019_lvl8

Course Code                                             Course  EOS    Mid
0         AL801    Software Design with Virtual Reality and Gaming  304  328.0
1         AL802               Software Design with Cloud Computing  301  306.0
2         AL803  Software Design with Mobile Apps and Connected...  309  337.0
3         AL805        Network Management and Cloud Infrastructure  329  442.0
4         AL810                                 Quantity Surveying  307  349.0
..          ...                                                ...  ...    ...
925       WD200                                     Arts (options)  221  296.0
926       WD210                       Software Systems Development  271  329.0
927       WD211                                 Creative Computing  275  322.0
928       WD212                    Recreation and Sport Management  274  311.0
929       WD230           Mechanical and Manufacturing Engineering  273  348.0

[926 rows x 4 columns]

In [268]:
# Need to Create a HEI code to get the College Name information. Use the first 2 Characters of the Course Code for this and create a new Column.  
df2019_lvl8['HEI Code'] = df2019_lvl8['Course Code'].str[:2]
df2019_lvl8

Course Code                                             Course  EOS  \
0         AL801    Software Design with Virtual Reality and Gaming  304   
1         AL802               Software Design with Cloud Computing  301   
2         AL803  Software Design with Mobile Apps and Connected...  309   
3         AL805        Network Management and Cloud Infrastructure  329   
4         AL810                                 Quantity Surveying  307   
..          ...                                                ...  ...   
925       WD200                                     Arts (options)  221   
926       WD210                       Software Systems Development  271   
927       WD211                                 Creative Computing  275   
928       WD212                    Recreation and Sport Management  274   
929       WD230           Mechanical and Manufacturing Engineering  273   

       Mid HEI Code  
0    328.0       AL  
1    306.0       AL  
2    337.0       AL  
3    442.0       AL  
4    349.0       AL  
..     ...      ...  
925  296.0       WD  
926  329.0       WD  
927  322.0       WD  
928  311.0       WD  
929  348.0       WD  

[926 rows x 5 columns]

In [269]:
# Temporary dataframe to get the HEI details and ISCED Category from the 2020 file. 
df=pd.merge(df2019_lvl8, df2020_lvl8[['Course Code','HEI','ISCED']], on='Course Code', how='left')
df

Course Code                                             Course  EOS  \
0         AL801    Software Design with Virtual Reality and Gaming  304   
1         AL802               Software Design with Cloud Computing  301   
2         AL803  Software Design with Mobile Apps and Connected...  309   
3         AL805        Network Management and Cloud Infrastructure  329   
4         AL810                                 Quantity Surveying  307   
..          ...                                                ...  ...   
921       WD200                                     Arts (options)  221   
922       WD210                       Software Systems Development  271   
923       WD211                                 Creative Computing  275   
924       WD212                    Recreation and Sport Management  274   
925       WD230           Mechanical and Manufacturing Engineering  273   

       Mid HEI Code                                HEI  \
0    328.0       AL    Athlone Institute of Technology   
1    306.0       AL    Athlone Institute of Technology   
2    337.0       AL    Athlone Institute of Technology   
3    442.0       AL    Athlone Institute of Technology   
4    349.0       AL    Athlone Institute of Technology   
..     ...      ...                                ...   
921  296.0       WD  Waterford Institute of Technology   
922  329.0       WD  Waterford Institute of Technology   
923  322.0       WD  Waterford Institute of Technology   
924  311.0       WD  Waterford Institute of Technology   
925  348.0       WD  Waterford Institute of Technology   

                                                 ISCED  
0    Information and Communication Technologies (ICTs)  
1    Information and Communication Technologies (ICTs)  
2    Information and Communication Technologies (ICTs)  
3    Information and Communication Technologies (ICTs)  
4                        Architecture and construction  
..                                                 ...  
921                                               Arts  
922  Information and Communication Technologies (ICTs)  
923  Information and Communication Technologies (ICTs)  
924                                  Personal services  
925                 Engineering and engineering trades  

[926 rows x 7 columns]

In [270]:
df2019_lvl8 =df[['Course Code', 'Course', 'HEI Code', 'HEI', 'ISCED', 'EOS']]
df2019_lvl8 = df2019_lvl8.rename(columns={'Course':'Course Title','EOS':'R1 Points'})
df2019_lvl8

Course Code                                       Course Title HEI Code  \
0         AL801    Software Design with Virtual Reality and Gaming       AL   
1         AL802               Software Design with Cloud Computing       AL   
2         AL803  Software Design with Mobile Apps and Connected...       AL   
3         AL805        Network Management and Cloud Infrastructure       AL   
4         AL810                                 Quantity Surveying       AL   
..          ...                                                ...      ...   
921       WD200                                     Arts (options)       WD   
922       WD210                       Software Systems Development       WD   
923       WD211                                 Creative Computing       WD   
924       WD212                    Recreation and Sport Management       WD   
925       WD230           Mechanical and Manufacturing Engineering       WD   

                                   HEI  \
0      Athlone Institute of Technology   
1      Athlone Institute of Technology   
2      Athlone Institute of Technology   
3      Athlone Institute of Technology   
4      Athlone Institute of Technology   
..                                 ...   
921  Waterford Institute of Technology   
922  Waterford Institute of Technology   
923  Waterford Institute of Technology   
924  Waterford Institute of Technology   
925  Waterford Institute of Technology   

                                                 ISCED R1 Points  
0    Information and Communication Technologies (ICTs)       304  
1    Information and Communication Technologies (ICTs)       301  
2    Information and Communication Technologies (ICTs)       309  
3    Information and Communication Technologies (ICTs)       329  
4                        Architecture and construction       307  
..                                                 ...       ...  
921                                               Arts       221  
922  Information and Communication Technologies (ICTs)       271  
923  Information and Communication Technologies (ICTs)       275  
924                                  Personal services       274  
925                 Engineering and engineering trades       273  

[926 rows x 6 columns]

In [271]:
df2019_lvl8['Year']=str(2019)
df2019_lvl8

Course Code                                       Course Title HEI Code  \
0         AL801    Software Design with Virtual Reality and Gaming       AL   
1         AL802               Software Design with Cloud Computing       AL   
2         AL803  Software Design with Mobile Apps and Connected...       AL   
3         AL805        Network Management and Cloud Infrastructure       AL   
4         AL810                                 Quantity Surveying       AL   
..          ...                                                ...      ...   
921       WD200                                     Arts (options)       WD   
922       WD210                       Software Systems Development       WD   
923       WD211                                 Creative Computing       WD   
924       WD212                    Recreation and Sport Management       WD   
925       WD230           Mechanical and Manufacturing Engineering       WD   

                                   HEI  \
0      Athlone Institute of Technology   
1      Athlone Institute of Technology   
2      Athlone Institute of Technology   
3      Athlone Institute of Technology   
4      Athlone Institute of Technology   
..                                 ...   
921  Waterford Institute of Technology   
922  Waterford Institute of Technology   
923  Waterford Institute of Technology   
924  Waterford Institute of Technology   
925  Waterford Institute of Technology   

                                                 ISCED R1 Points  Year  
0    Information and Communication Technologies (ICTs)       304  2019  
1    Information and Communication Technologies (ICTs)       301  2019  
2    Information and Communication Technologies (ICTs)       309  2019  
3    Information and Communication Technologies (ICTs)       329  2019  
4                        Architecture and construction       307  2019  
..                                                 ...       ...   ...  
921                                               Arts       221  2019  
922  Information and Communication Technologies (ICTs)       271  2019  
923  Information and Communication Technologies (ICTs)       275  2019  
924                                  Personal services       274  2019  
925                 Engineering and engineering trades       273  2019  

[926 rows x 7 columns]

In [280]:
# Check the R1 points, have a look at any that dont begin with and contain only numbers from 0 - 9.
df2019_lvl8[df2019_lvl8['R1 Points'].str.contains('^\d')==False]

Course Code                                       Course Title HEI Code  \
23        AL861             Animation and Illustration (portfolio)       AL   
24        AL863             Graphic and Digital Design (portfolio)       AL   
29        CW038                           Art (portfolio, Wexford)       CW   
56        CW858  Sports Management and Coaching (options, portf...       CW   
70        CR121                  Music at CIT Cork School of Music       CR   
..          ...                                                ...      ...   
825       MH802    Community and Youth Work (part-time, in service       MH   
826       MH803  Local Studies or Community Studies (part-time,...       MH   
844       SG244                                           Fine Art       SG   
886       TL889  Counselling with Addiction - Mature applicants...       TL   
895       WD027                                              Music       WD   

                                   HEI                            ISCED  \
23     Athlone Institute of Technology                             Arts   
24     Athlone Institute of Technology                             Arts   
29     Institute of Technology, Carlow                             Arts   
56     Institute of Technology, Carlow                Personal services   
70        Cork Institute of Technology                             Arts   
..                                 ...                              ...   
825                Maynooth University                          Welfare   
826                Maynooth University  Social and behavioural sciences   
844     Institute of Technology, Sligo                             Arts   
886                                NaN                              NaN   
895  Waterford Institute of Technology                             Arts   

    R1 Points  Year  
23       #615  2019  
24       #703  2019  
29       #700  2019  
56       #700  2019  
70       #633  2019  
..        ...   ...  
825         #  2019  
826         #  2019  
844      #355  2019  
886         #  2019  
895      #290  2019  

[95 rows x 7 columns]

In [272]:
# Save the 2019 file. 
path = cwd + '/CAO_Data'+'/cao2019_' + now.strftime('%Y%m%d_%H%M%S')+'.csv'
df2019_lvl8.to_csv(path, index=False)

In [275]:
# Join the 2019, 2020 and 2021 files.
allcourses = pd.concat([df2021_lvl8,df2020_lvl8, df2019_lvl8])
allcourses.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2941 entries, 29 to 925
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Course Code   2941 non-null   object
 1   Course Title  2941 non-null   object
 2   HEI Code      2941 non-null   object
 3   HEI           2792 non-null   object
 4   ISCED         2792 non-null   object
 5   R1 Points     2941 non-null   object
 6   Year          2941 non-null   object
dtypes: object(7)
memory usage: 183.8+ KB


In the consolidated file look like there are quite a few records missing the HEI, 2941 HEI codes but 2792 HEI.
111 of these relate to the Technological University of Dublin. 

In [284]:
# This is the list courses that have moved from the BN, DT, TA course codes in 2019 to the TU codes in 2020 and 2021
tulist=['BN','DT','TA']
result=df2019_lvl8[df2019_lvl8['HEI Code'].isin(tulist)]
result

Course Code                                     Course Title HEI Code  \
265       DT001                                   Product Design       DT   
266       DT028              Transport Operations and Technology       DT   
267       DT035        Electrical Services and Energy Management       DT   
268       DT066                      Engineering (General Entry)       DT   
269       DT101                Architecture (test and interview)       DT   
..          ...                                              ...      ...   
407       TA323                     Computing with IT Management       TA   
408       TA326                        DNA and Forensic Analysis       TA   
409       TA327                        Sports Science and Health       TA   
410       TA328  Computing with Language (French/German/Spanish)       TA   
411       TA329                    Computing with Data Analytics       TA   

     HEI ISCED R1 Points  Year  
265  NaN   NaN       433  2019  
266  NaN   NaN       279  2019  
267  NaN   NaN       293  2019  
268  NaN   NaN       400  2019  
269  NaN   NaN      #618  2019  
..   ...   ...       ...   ...  
407  NaN   NaN       243  2019  
408  NaN   NaN       300  2019  
409  NaN   NaN       341  2019  
410  NaN   NaN       218  2019  
411  NaN   NaN       218  2019  

[111 rows x 7 columns]

In [311]:
result1=pd.merge(result, df2020_lvl8[['Course Code','Course Title','HEI Code','HEI','ISCED', 'R1 Points', 'Year']], on='Course Title', how='left')
result1=result1[result1['HEI Code_y']=='TU']
result1

Course Code_x                                     Course Title HEI Code_x  \
1           DT001                                   Product Design         DT   
2           DT028              Transport Operations and Technology         DT   
3           DT035        Electrical Services and Energy Management         DT   
4           DT066                      Engineering (General Entry)         DT   
6           DT106            Planning and Environmental Management         DT   
..            ...                                              ...        ...   
222         TA323                     Computing with IT Management         TA   
223         TA326                        DNA and Forensic Analysis         TA   
224         TA327                        Sports Science and Health         TA   
225         TA328  Computing with Language (French/German/Spanish)         TA   
226         TA329                    Computing with Data Analytics         TA   

    HEI_x ISCED_x R1 Points_x Year_x Course Code_y HEI Code_y  \
1     NaN     NaN         433   2019         TU811         TU   
2     NaN     NaN         279   2019         TU814         TU   
3     NaN     NaN         293   2019         TU802         TU   
4     NaN     NaN         400   2019         TU805         TU   
6     NaN     NaN         312   2019         TU835         TU   
..    ...     ...         ...    ...           ...        ...   
222   NaN     NaN         243   2019         TU865         TU   
223   NaN     NaN         300   2019         TU864         TU   
224   NaN     NaN         341   2019         TU936         TU   
225   NaN     NaN         218   2019         TU861         TU   
226   NaN     NaN         218   2019         TU862         TU   

                               HEI_y  \
1    Technological University Dublin   
2    Technological University Dublin   
3    Technological University Dublin   
4    Technological University Dublin   
6    Technological University Dublin   
..                               ...   
222  Technological University Dublin   
223  Technological University Dublin   
224  Technological University Dublin   
225  Technological University Dublin   
226  Technological University Dublin   

                                               ISCED_y R1 Points_y Year_y  
1                                                 Arts         411   2020  
2                                   Transport services         285   2020  
3                   Engineering and engineering trades         317   2020  
4                   Engineering and engineering trades         402   2020  
6                        Architecture and construction         313   2020  
..                                                 ...         ...    ...  
222  Information and Communication Technologies (ICTs)         226   2020  
223                    Biological and related sciences         400   2020  
224                    Biological and related sciences         410   2020  
225  Information and Communication Technologies (ICTs)         308   2020  
226  Information and Communication Technologies (ICTs)         226   2020  

[118 rows x 13 columns]

In [313]:
ids=result1['Course Code_x']
result1[ids.isin(ids[ids.duplicated()])]

Course Code_x                        Course Title HEI Code_x HEI_x  \
67          DT366              Accounting and Finance         DT   NaN   
68          DT366              Accounting and Finance         DT   NaN   
69          DT366              Accounting and Finance         DT   NaN   
131         BN110              International Business         BN   NaN   
132         BN110              International Business         BN   NaN   
136         BN112              Creative Digital Media         BN   NaN   
137         BN112              Creative Digital Media         BN   NaN   
148         BN114              Accounting and Finance         BN   NaN   
149         BN114              Accounting and Finance         BN   NaN   
150         BN114              Accounting and Finance         BN   NaN   
156         BN118  Early Childhood Care and Education         BN   NaN   
157         BN118  Early Childhood Care and Education         BN   NaN   
167         TA022              Creative Digital Media         TA   NaN   
168         TA022              Creative Digital Media         TA   NaN   
173         TA024  Early Childhood Care and Education         TA   NaN   
174         TA024  Early Childhood Care and Education         TA   NaN   
190         TA121              Accounting and Finance         TA   NaN   
191         TA121              Accounting and Finance         TA   NaN   
192         TA121              Accounting and Finance         TA   NaN   
197         TA125              International Business         TA   NaN   
198         TA125              International Business         TA   NaN   

    ISCED_x R1 Points_x Year_x Course Code_y HEI Code_y  \
67      NaN         432   2019         TU909         TU   
68      NaN         432   2019         TU910         TU   
69      NaN         432   2019         TU911         TU   
131     NaN         253   2019         TU915         TU   
132     NaN         253   2019         TU916         TU   
136     NaN         280   2019         TU981         TU   
137     NaN         280   2019         TU982         TU   
148     NaN         280   2019         TU909         TU   
149     NaN         280   2019         TU910         TU   
150     NaN         280   2019         TU911         TU   
156     NaN         290   2019         TU993         TU   
157     NaN         290   2019         TU994         TU   
167     NaN         260   2019         TU981         TU   
168     NaN         260   2019         TU982         TU   
173     NaN         217   2019         TU993         TU   
174     NaN         217   2019         TU994         TU   
190     NaN         202   2019         TU909         TU   
191     NaN         202   2019         TU910         TU   
192     NaN         202   2019         TU911         TU   
197     NaN         229   2019         TU915         TU   
198     NaN         229   2019         TU916         TU   

                               HEI_y                      ISCED_y R1 Points_y  \
67   Technological University Dublin  Business and administration         285   
68   Technological University Dublin  Business and administration         444   
69   Technological University Dublin  Business and administration         209   
131  Technological University Dublin  Business and administration         252   
132  Technological University Dublin  Business and administration         235   
136  Technological University Dublin                         Arts         280   
137  Technological University Dublin                         Arts         320   
148  Technological University Dublin  Business and administration         285   
149  Technological University Dublin  Business and administration         444   
150  Technological University Dublin  Business and administration         209   
156  Technological University Dublin                      Welfare         270   
157  Technological University Dublin                      Welfare         321   
167  Technological Univ

In [277]:
# Check which course codes are duplicated.
#allcourses[allcourses.duplicated(subset=['Course Code'])]

In [ ]:
# Only show the unique course codes. 
allcourses.drop_duplicates(subset=['ccode'], inplace=True, ignore_index=True)
allcourses

In [ ]:
allcourses.set_index('ccode').join(df2020.set_index('COURSE CODE2'))

In [ ]:
testjoin=allcourses.merge(df2020[['COURSE CODE2','R1 POINTS']],how='left',left_on='ccode', right_on='COURSE CODE2')
testjoin